### DISCUSSION TITLE
Динамический топ за месяц через Lastfm.getCustomTop

### DISCUSSION THREAD

**Discussion by amd64fox** [Chimildic/goofy#46] [Помощь с алгоритмом] [answered]
Написал такой код :

```
var date = new Date(), y = date.getFullYear(), m = date.getMonth();
var firstDay = new Date(y, m, 1).toLocaleDateString("ru-RU", { timeZone: "Europe/Moscow"})
var lastDay = new Date(y, m + 1, 0).toLocaleDateString("ru-RU", { timeZone: "Europe/Moscow"})

console.log(firstDay);
console.log(lastDay);
```
Для этой функции :

```
let tracks = Lastfm.getCustomTop({
    user: '',
    from: firstDay,
    to: lastDay,
    limit: 50 
});
```
1) Не могу поставить вместо точек тире, сейчас вот так `01.02.2021`, а нужно так `01-02-2021`
2) Цель получать топ треки отдельно для каждого месяцы. Но как это проверить будет ли это работать когда сменится месяц ?, что бы не ждать 28 числа)

**Comment by Chimildic** [comment]
Если нужен топ на месяц, можно использовать [getTopTracks](https://chimildic.github.io/goofy/#/func?id=gettoptracks-1) с периодом `1month` и триггером на конец месяца.

`getCustomTop` нужен прежде всего, чтобы заглядывать дальше одного года назад. Поскольку Last.fm API дает максимум `12month`.

**Reply by amd64fox** [reply]
функция `getTopTracks` с периодом `1month` работает не совсем так как нужно, например если наступает 1-ое число нового месяца, то она считает это первое число + 29 дней старого месяца, а мне нужно что бы посчитался только первый день нового, ну и так далее.

**Reply by Chimildic** [reply]
То есть обновлялось не раз в месяц, а каждый день?

**Reply by amd64fox** [reply]
у меня этот скрипт выполняется каждый день да

**Comment by Chimildic** [comment]
- Поскольку триггер будет работать каждый день, можно не вычислять последнее число текущего месяца. 
- Внутри `getCustomTop` строка-даты проходит через `new Date`, важен формат. Дату вида `08.02.2021` конструктор не понимает (локаль `ru-RU`). Но понимает `2/8/2021` (локаль `en-US`).

```js
function createTopMonth() {
    let period = makeMonthPeriod();
    let tracks = Lastfm.getCustomTop({
        user: '',
        from: period.from,
        to: period.to,
        limit: 50,
    });
    Playlist.saveWithReplace({
        id: '',
        tracks: tracks,
    });

    function makeMonthPeriod() {
        let fromDay = new Date();
        fromDay.setDate(1);
        return {
            from: fromDay.toLocaleDateString('en-US', { timeZone: 'Europe/Moscow' }),
            to: new Date().toLocaleDateString('en-US', { timeZone: 'Europe/Moscow' }),
        };
    }
}
```

**Reply by vlalog** [reply]
А может изменить входные параметры на объект Date и уже в теле самой функции преобразовать в нужный формат 🤔

**Reply by vlalog** [reply]
или проверять тип "строка"/"Date object"

**Reply by amd64fox** [reply]
Отлично, первый вариант понял, работает, спасибо,  осталось проверить будет ли он каждый, новый месяц так работать. А второй вариант мне из-за незнания программирования без примера очень сложно понять, я на код иногда смотрю как папуас на Миклухо-Маклая), но все равно спасибо.

**Comment by Chimildic** [comment, marked as answer]
Добавил проверку типа. Если обновите библиотеку, сможете использовать следующий код. Принцип тот же: первое и текущее число месяца.
```js
  let tracks = Lastfm.getCustomTop({
    user: '',
    from: new Date().setDate(1),
    to: new Date(),
    limit: 50
  });
```

**Reply by amd64fox** [reply]
Обновился, спасибо !)
